doc/how-to-analyze-haskell-code-performance.rst

```diff
@@ -0,0 +1,126 @@
+How to analyze Haskell performance
+==================================
+
+When a Haskell application is slow or uses too much memory,
+Cabal and `GHC <https://downloads.haskell.org/ghc/latest/docs/users_guide/>`__
+can help you understand why.
+
+The main steps are to
+
+1. let GHC insert performance measuring code into your application,
+2. run the application to collect a performance report and
+3. visualize that report.
+
+The process of inserting performance measuring code and collecting performance information
+is called "profiling". The main work is done by GHC while Cabal only acts
+as a build configuration interface and passes corresponding compiler flags to GHC.
+
+Profiling CPU performance
+-------------------------
+
+The first step to build your application, e.g. ``my-app``, with profiling enabled, and
+the second step to run it to collect a report, can be done with a single ``cabal run`` command:
+
+.. code-block:: console
+
+      $ cabal run --enable-profiling --profiling-detail=late my-app -- +RTS -pj -RTS
+      <program runs and finishes>
+
+Finally, a profiling JSON report is written to a ``<app-name>.prof`` file,
+i.e. ``my-app.prof``, in the current directory.
+Load the profiling report file  ``my-app.prof`` into a visualizer
+and look for performance bottlenecks. One popular open-source
+`flame graph <https://www.brendangregg.com/flamegraphs.html>`__
+visualizer is
+`Speedscope <https://speedscope.app>`__,
+which runs in the browser and can open this JSON file directly.
+See the
+`Haskell Optimization Handbook <https://haskell.foundation/hs-opt-handbook.github.io>`__
+on how to optimize your code based on the profiling results afterwards.
+
+The ``cabal run`` command above is essentially a shorthand for
+
+.. code-block:: console
+
+    $ cabal build --enable-profiling --profiling-detail=late my-app
+    $ cabal list-bin my-app
+    /path/to/my-app
+    $ /path/to/my-app +RTS -pj -RTS
+    <program runs and finishes>
```

Comment:
As said, I would mention this first and then the compressed command.

Reply:
Done